### PR TITLE
Added field privacy for #86

### PR DESCRIPTION
**Pull request by TedDriggs**
**Open Question**: Should validation disallow setter being less public than field?

### CODE REVIEW

File: derive_builder/src/options/field_mode.rs [hunk 1]
```diff
@@ -125,23 +126,26 @@ impl From<OptionsBuilder<FieldMode>> for FieldOptions {
                     },
                     _ => syn::Ident::new(field_ident.clone()),
                 }});
+                
+        let setter_vis = b.setter_vis.unwrap_or(syn::Visibility::Public);
 
         FieldOptions {
             setter_enabled: b.setter_enabled.unwrap_or(true),
-            builder_pattern: b.builder_pattern.clone().unwrap_or_default(),
```
Comment (TedDriggs) [inline derive_builder/src/options/field_mode.rs]: @colin-kiegel I wasn't sure where all these `clone()` calls came from, but removing them didn't break any of the tests. There may not be any efficiency impact, but it's cleaner now :)

File: derive_builder/src/options/mod.rs [hunk 2]
```diff
@@ -228,7 +234,7 @@ impl<Mode> OptionsBuilder<Mode> where
                 self.setter_public(false)
             },
             "setter" => {
-                // setter implicitly enabled
+        // setter implicitly enabled
```
Comment (colin-kiegel) [inline derive_builder/src/options/mod.rs]: hm, that indentation looks strange. It's not aligned with the next line anymore.

File: derive_builder/src/options/mod.rs [hunk 3]
```diff
@@ -293,32 +299,44 @@ impl<Mode> OptionsBuilder<Mode> where
...
-        ident: &syn::Ident,
-        nested: &[syn::NestedMetaItem]
-    ) {
+    fn parse_builder_options_list(&mut self, ident: &syn::Ident, nested: &[syn::NestedMetaItem]) {
         trace!("Parsing list `{}({:?})`", ident.as_ref(), nested);
         match ident.as_ref() {
             "setter" => {
                 self.parse_setter_options(nested);
-                // setter implicitly enabled
+        // setter implicitly enabled
```
Comment (colin-kiegel) [inline derive_builder/src/options/mod.rs]: new indentation looks strange.

File: derive_builder_core/src/builder_field.rs [hunk 4]
```diff
@@ -33,12 +33,14 @@ pub struct BuilderField<'a> {
...
-    /// Wether the builder implements a setter for this field.
+    /// Whether the builder implements a setter for this field.
     ///
     /// Note: We will fallback to `PhantomData` if the setter is disabled
     ///       to hack around issues with unused generic type parameters - at least for now.
     pub setter_enabled: bool,
     /// Visibility of this builder field, e.g. `syn::Visibility::Public`.
+    pub field_visibility: &'a syn::Visibility,
+    /// Visibility of this field's setter, e.g. `syn::Visibility::Public`.
     pub setter_visibility: &'a syn::Visibility,
```
Comment (colin-kiegel) [inline derive_builder_core/src/builder_field.rs]: this old line can now be removed

**Comment by colin-kiegel** [comment]
Oh, please feel free to add yourself to the list of authors, too. ;-)

File: derive_builder_core/src/builder_field.rs (hunk 4 above)
Comment (TedDriggs) [inline derive_builder_core/src/builder_field.rs]: Which old line? I ended up using both the field and setter visibility.

**Comment by TedDriggs** [comment]
> Would you agree, that we should change the default? If yes, I'd again prefer to phase-in this change via deprecation messages.

100% agree; I'll try to get this added in. Should there be a crate feature for `private_fields`, like there was for `struct_default`? We'd squelch the warning if the crate feature was enabled OR the struct explicitly sets field visibility at the struct level.

I'll also add tests for the next commit.

File: derive_builder/src/options/mod.rs (hunk 2 above)
Comment (TedDriggs) [inline derive_builder/src/options/mod.rs]: I'll fix this, and see if an issue needs to get filed against the formatter.

File: derive_builder_core/src/builder_field.rs (hunk 4 above)
Comment (colin-kiegel) [inline derive_builder_core/src/builder_field.rs]: `BuilderField::setter_visibility` can be removed. It's only usage is one initialisation in the `default_builder_field!` macro.

File: derive_builder/src/options/struct_mode.rs [hunk 5]
```diff
@@ -167,7 +167,21 @@ impl OptionsBuilderMode for StructMode {
...
+        // present a compilation warning.
+        if !cfg!(feature = "private_fields") && b.field_vis.is_none() {
+            let where_diagnostics = b.where_diagnostics();
+            b.mode.push_deprecation_note(format!(
+                "Builder fields will be private by default starting in the next version. \
+                (see https://github.com/colin-kiegel/rust-derive-builder/issues/86 for \
+                more details). To squelch this message and adopt the new behavior now, \
+                compile `derive_builder` with `--features \"private_fields\"` or add \
+                `field(<vis>)` to the builder attribute on the struct. (Found {})",
+                where_diagnostics));
```
Comment (colin-kiegel) [inline derive_builder/src/options/struct_mode.rs]: :+1: :-)

File: .travis.yml [hunk 6]
```diff
@@ -8,7 +8,7 @@ rust:
 matrix:
   include:
     - rust: 1.15.0
-      env: JOB=build CARGO_FEATURES="skeptic_tests struct_default"
+      env: JOB=build CARGO_FEATURES="skeptic_tests struct_default private_fields"
```
Comment (TedDriggs) [inline .travis.yml]: Done.

**Comment by TedDriggs** [comment]
> What was that fourth-last commit?

Probably a mishap doing a merge or rebase. I'm still a bit new to Git so I sometimes end up with errant commits in my history.

> remove `BuilderField::setter_visibility`

I left this in on purpose: It allows someone to have a public setter while still having the backer field be private.

> I'd like to have an integration test - and we don't need the example. I generally prefer integration tests over examples - since we already have good documentation.

I've moved it to a test.

**Comment by colin-kiegel** [comment]
>> remove `BuilderField::setter_visibility`
> 
> I left this in on purpose: It allows someone to have a public setter while still having the backer field be private.

Ok, I think we were really talking about something different. When you look at my [last commit](https://github.com/colin-kiegel/rust-derive-builder/pull/87/commits/f63ee34381188570bf546afc70f9e37d973359a1#diff-40ccb9389447ef5f7532c1bcfd9aba3c), you can see what I meant by dead code. All tests still pass after this removal. :-)

**Comment by colin-kiegel** [comment]
@TedDriggs I am going to publish a new release later today.

Please note that I will be on vacation for a _couple of weeks_ starting on sunday. I think I will be able to do reviews via github during this time - but I may not be able to publish new releases.

**Comment by TedDriggs** [comment]
Have a great vacation! I'll find another library to bother during that time :)

**Pull request merged**